pkg/cmd/registry/artifact/crud/create/create.go

```diff
@@ -0,0 +1,200 @@
+	Logger     func() (logging.Logger, error)
+	localizer  localize.Localizer
+}
+
+var longDescription = `
+Creates a new artifact by posting the artifact content to the registry.
+
+Artifacts can be typically in JSON format for most of the supported types, but may be in another format for a few (for example, PROTOBUF).
+The registry attempts to figure out what kind of artifact is being added from the following supported list:
+
+- Avro (AVRO)
+- Protobuf (PROTOBUF)
+- JSON Schema (JSON)
+- Kafka Connect (KCONNECT)
+- OpenAPI (OPENAPI)
+- AsyncAPI (ASYNCAPI)
+- GraphQL (GRAPHQL)
+- Web Services Description Language (WSDL)
+- XML Schema (XSD)
+
+An artifact is created using the content provided in the body of the request.  
+This content is created under a unique artifact ID that can be provided by user.
+If not provided in the request, the server generates a unique ID for the artifact. 
+It is typically recommended that callers provide the ID, because this is a meaningful identifier, and for most use cases should be supplied by the caller.
+If an artifact with the provided artifact ID already exists command will fail with error.
+
+
+When --group parameter is missing the command will create a new artifact under the "default" group.
+when --registry is missing the command will create a new artifact for currently active service registry (visible in rhoas service-registry describe)
+`
+
+func NewCreateCommand(f *factory.Factory) *cobra.Command {
+	opts := &Options{
+		IO:         f.IOStreams,
+		Config:     f.Config,
+		Connection: f.Connection,
+		Logger:     f.Logger,
+		localizer:  f.Localizer,
+	}
+
+	cmd := &cobra.Command{
+		Use:   "create",
+		Short: "Creates new artifact from file or standard input",
+		Long:  longDescription,
+		Example: `
+# Create an artifact in default group
+rhoas service-registry artifact create my-artifact.json
+
+# Create an artifact with specified type
+rhoas service-registry artifact create --type=JSON my-artifact.json
+		`,
+		Args: cobra.RangeArgs(0, 1),
+		RunE: func(cmd *cobra.Command, args []string) error {
+			validOutputFormats := flagutil.ValidOutputFormats
+			if opts.outputFormat != "" && !flagutil.IsValidInput(opts.outputFormat, validOutputFormats...) {
+				return flag.InvalidValueError("output", opts.outputFormat, validOutputFormats...)
+			}
+
+			if len(args) > 0 {
+				opts.file = args[0]
+			}
+
+			if opts.registryID != "" {
+				return runCreate(opts)
+			}
+
+			cfg, err := opts.Config.Load()
+			if err != nil {
+				return err
+			}
+
+			if opts.artifactType != "" {
+				if _, err = registryinstanceclient.NewArtifactTypeFromValue(opts.artifactType); err != nil {
+					return errors.New("Invalid artifact type. Please use one of following values: " + util.GetAllowedArtifactTypeEnumValuesAsString())
+				}
+			}
+
+			if !cfg.HasServiceRegistry() {
+				return errors.New("No service Registry selected. Use 'rhoas service-registry use' use to select your registry")
```

Comment:
**me thinking** When this is running in the independent upstream CLI, what will the error message say, because here it directly refers to `rhoas` but that would not always be so.

Reply:
Good point!

Reply:
Special error types would be perhaps be needed which are scoped to the data plane CLI.

From data plane CLI, keep it simple with no reference to what command you should use:

```go
var NoRegistrySelectedError = errors.New("no service registry selected")

return NoRegistrySelectedError
```

When interpreting the error, add additional info:

```go
if rerr, ok := err.(*NoRegistrySelectedError); ok {
  return fmt.Errorf("%w: use 'rhoas service-registry use' to select a registry or pass the `--registry-id` flag")
}
```

That is a simplistic approach which does not scale very well, but perhaps some kind of abstraction would be helpful to match errors.

Reply:
Good idea!

Reply:
Created dedicated issue as this is not instantly actionable in PR. I will discuss that with the team ( CC @ankithans )